go/src/koding/kites/kloud/provider/marathon/stack.go

```diff
@@ -0,0 +1,406 @@
+	t := s.Builder.Template
+
+	var resource struct {
+		MarathonApp map[string]map[string]interface{} `hcl:"marathon_app"`
+	}
+
+	if err := t.DecodeResource(&resource); err != nil {
+		return nil, err
+	}
+
+	if len(resource.MarathonApp) == 0 {
+		return nil, errors.New("applications are empty")
+	}
+
+	for _, app := range resource.MarathonApp {
+		originalAppID := s.convertInstancesToGroup(app)
+
+		labels, err := s.injectEntrypoint(app, originalAppID)
+		if err != nil {
+			return nil, err
+		}
+
+		s.injectFetchEntrypoints(app, len(labels))
+		s.injectHealthChecks(app)
+
+		if err := s.injectMetadata(app, labels); err != nil {
+			return nil, err
+		}
+	}
+
+	t.Resource["marathon_app"] = resource.MarathonApp
+
+	err := t.ShadowVariables("FORBIDDEN", "marathon_basic_auth_user", "marathon_basic_auth_password")
+	if err != nil {
+		return nil, errors.New("marathon: error shadowing: " + err.Error())
+	}
+
+	if err := t.Flush(); err != nil {
+		return nil, errors.New("marathom: error flushing template: " + err.Error())
+	}
+
+	content, err := t.JsonOutput()
+	if err != nil {
+		return nil, err
+	}
+
+	return &stack.Template{
+		Content: content,
+	}, nil
+
+	return nil, nil
+}
+
+// convertInstancesToGroup converts instances property to a count one.
+//
+// Since Marathon does not support instance indexing it's not possible
+// to assign unique metadata for each of the instace, thus making such
+// stack unusable for Koding. Relevant issue:
+//
+//   https://github.com/mesosphere/marathon/issues/1242
+//
+// What we do instead is we convert multiple instances of an application to
+// an application group as a workaround.
+func (s *Stack) convertInstancesToGroup(app map[string]interface{}) (originalAppID string) {
+	instances, ok := app["instances"].(int)
+	if ok {
+		delete(app, "instances")
+	} else {
+		instances = 1
+	}
+
+	count, ok := app["count"].(int)
+	if !ok {
+		count = 1
+	}
+
+	count *= instances
+
+	app["count"] = count
+
+	// Each app within group must have unique name.
+	appID, ok := app["app_id"].(string)
+
+	if ok && appID != "" && count > 1 {
+		app["app_id"] = path.Join(appID, path.Base(appID)+"-${count.index + 1}")
+	}
+
+	return appID
+}
+
+var ErrIncompatibleEntrypoint = errors.New(`marathon: setting "args" argument conflicts with Koding entrypoint injected into each container. Please use "cmd" argument instead.`)
+
+// injectEntrypoint injects an entrypoint, which is responsible for installing
+// klient before running container's command.
+//
+// The entrypoint is injected in a twofold manner:
+//
+//   - if "cmd" argument is used, it's prefixed with an entrypoint.N.sh script
+//   - if container's default command (the one from Dockerfile) is used,
+//     the container's entrypoint (by default /bin/sh) is replaced
+//     with the Koding one
+//
+func (s *Stack) injectEntrypoint(app map[string]interface{}, originalAppID string) (labels []string, err error) {
```

Comment:
Functional tests, yes. The integration tests would be added with kd. Yes, these tests have pretty much good coverage.

There are a lots of TODOs in this provider - e.g. if Mesos add INSTANCE_ID we will get rid of converting instances to groups. I chose not to add unittests that test against non-exported API in order to not have to maintain tests each time this private API changes (e.g. convert* is removed). And sometimes it's impossible to refactor existing code with having tests passing, since a lot of unittests in our codebase call unexported methods directly.